[WebAssembly SIMD] Support floating-point absolute value, negate, min, max, and square-root on Intel

ddegazio opened the pull request:
#### 3a6ea5eb8b9b6395b6ee4fe11afb07861cea519c
<pre>
[WebAssembly SIMD] Support floating-point absolute value, negate, min, max, and square-root on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=249275">https://bugs.webkit.org/show_bug.cgi?id=249275</a>
rdar://103326365

Reviewed by Yusuke Suzuki.

Adds support for missing floating-point opcodes to our WebAssembly SIMD support on Intel, with some
fixes to bugs that came up running the spec tests.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVectorUnordered):
(JSC::MacroAssemblerX86_64::vectorMax):
(JSC::MacroAssemblerX86_64::vectorMin):
(JSC::MacroAssemblerX86_64::vectorPmin):
(JSC::MacroAssemblerX86_64::vectorPmax):
(JSC::MacroAssemblerX86_64::vectorTruncSatSignedFloat64):
(JSC::MacroAssemblerX86_64::vectorSqrt):
(JSC::MacroAssemblerX86_64::vectorSshr8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
(JSC::MacroAssemblerX86_64::vectorSshr):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vmaxps_rrr):
(JSC::X86Assembler::vmaxpd_rrr):
(JSC::X86Assembler::vminps_rrr):
(JSC::X86Assembler::vminpd_rrr):
(JSC::X86Assembler::vsqrtps_rr):
(JSC::X86Assembler::vsqrtpd_rr):
(JSC::X86Assembler::vminpd_mrr):
(JSC::X86Assembler::vpsrlw_i8rr):
(JSC::X86Assembler::vpsrld_i8rr):
(JSC::X86Assembler::vpsrlq_i8rr):
(JSC::X86Assembler::vpsraw_i8rr):
(JSC::X86Assembler::vpsrad_i8rr):
(JSC::X86Assembler::vcmpeqpd_rrr): Deleted.
(JSC::X86Assembler::vpsraq_rrr): Deleted.
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDShift):

Canonical link: <a href="https://commits.webkit.org/257880@main">https://commits.webkit.org/257880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00812d115d7ee3ab6994b610cf33cc31bcc7cdb3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100264 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109588 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169815 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10335 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92680 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107475 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106040 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7822 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91101 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34499 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22491 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90827 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3185 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24010 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86818 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/629 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3166 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29079 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9293 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43500 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89698 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5418 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4994 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20051 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3128,6 +3133,34 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/minpd
+        // VEX.128.66.0F.WIG 5D /r VMINPD xmm1, xmm2, xmm3/m128
+        m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_66, OP2_MINPD_VpdWpd, (RegisterID)xmm1, (RegisterID)xmm3, (RegisterID)xmm2);
+    }
```
Let's move these functions to AVX section. In X86Assembler, we are consolidating all AVX functions in one place.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5008,6 +5041,16 @@ class X86Assembler {
     }
 
+    void vsqrtps_rr(XMMRegisterID src, XMMRegisterID dest)
+    {
```
Let's add the comment aligned to the other AVX functions, it is,
// URL
// Encoding
// Operand interpretation

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3128,6 +3133,34 @@ class X86Assembler {
+    void vminpd_rrr(XMMRegisterID xmm3, XMMRegisterID xmm2, XMMRegisterID xmm1)
+    {
+        // https://www.felixcloutier.com/x86/minpd
+        // VEX.128.66.0F.WIG 5D /r VMINPD xmm1, xmm2, xmm3/m128
```
Let's add operand interpretation table.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3128,6 +3133,34 @@ class X86Assembler {
+    void vminps_rrr(XMMRegisterID xmm3, XMMRegisterID xmm2, XMMRegisterID xmm1)
+    {
+        // https://www.felixcloutier.com/x86/minps
+        // VEX.128.NP.0F.WIG 5D /r VMINPS xmm1, xmm2, xmm3/m128
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3128,6 +3133,34 @@ class X86Assembler {
+    void vmaxpd_rrr(XMMRegisterID xmm3, XMMRegisterID xmm2, XMMRegisterID xmm1)
+    {
+        // https://www.felixcloutier.com/x86/maxpd
+        // VEX.128.66.0F.WIG 5F /r VMAXPD xmm1, xmm2, xmm3/m128
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3128,6 +3133,34 @@ class X86Assembler {
+    void vmaxps_rrr(XMMRegisterID xmm3, XMMRegisterID xmm2, XMMRegisterID xmm1)
+    {
+        // https://www.felixcloutier.com/x86/maxps
+        // VEX.128.NP.0F.WIG 5F /r VMAXPS xmm1, xmm2, xmm3/m128
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5008,6 +5041,16 @@ class X86Assembler {
+    }
+
+    void vsqrtpd_rr(XMMRegisterID src, XMMRegisterID dest)
+    {
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5153,10 +5180,44 @@ class X86Assembler {
-        // https://www.felixcloutier.com/x86/cmppd
-        vcmppd_rrr(0, xmm3, xmm2, xmm1);
+        // https://www.felixcloutier.com/x86/psrlw:psrld:psrlq
+        // VEX.128.66.0F.WIG 71 /2 ib VPSRLW xmm1, xmm2, imm8
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5153,10 +5180,44 @@ class X86Assembler {
+    void vpsrld_i8rr(uint8_t shift, XMMRegisterID src, XMMRegisterID dst)
+    {
+        // https://www.felixcloutier.com/x86/psrlw:psrld:psrlq
+        // VEX.128.66.0F.WIG 72 /2 ib VPSRLD xmm1, xmm2, imm8
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5153,10 +5180,44 @@ class X86Assembler {
+    void vpsrlq_i8rr(uint8_t shift, XMMRegisterID src, XMMRegisterID dst)
+    {
+        // https://www.felixcloutier.com/x86/psrlw:psrld:psrlq
+        // VEX.128.66.0F.WIG 73 /2 ib VPSRLQ xmm1, xmm2, imm8
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5153,10 +5180,44 @@ class X86Assembler {
+    void vpsraw_i8rr(uint8_t shift, XMMRegisterID src, XMMRegisterID dst)
+    {
+        // https://www.felixcloutier.com/x86/psraw:psrad:psraq
+        // VEX.128.66.0F.WIG 71 /4 ib VPSRAW xmm1, xmm2, imm8
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5153,10 +5180,44 @@ class X86Assembler {
+    void vpsrad_i8rr(uint8_t shift, XMMRegisterID src, XMMRegisterID dst)
+    {
+        // https://www.felixcloutier.com/x86/psraw:psrad:psraq
+        // VEX.128.66.0F.WIG 72 /4 ib VPSRAD xmm1, xmm2, imm8
```
Ditto.

Constellation reviewed (changes requested):
Commented about how we define AVX functions in X86Assembler. Let's make them consistent to the rest of definitions.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2343,6 +2343,19 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        using PackedCompareCondition = X86Assembler::PackedCompareCondition;
+
+        if (simdInfo.lane == SIMDLane::f32x4)
+            m_assembler.vcmpps_rrr(PackedCompareCondition::Unordered, left, right, dest);
```
Isn't it opposite? (right, left)

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2343,6 +2343,19 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        if (simdInfo.lane == SIMDLane::f32x4)
+            m_assembler.vcmpps_rrr(PackedCompareCondition::Unordered, left, right, dest);
+        else
+            m_assembler.vcmppd_rrr(PackedCompareCondition::Unordered, left, right, dest);
```
Ditto

Constellation reviewed (approved):
r=me

ddegazio committed:
[WebAssembly SIMD] Support floating-point absolute value, negate, min, max, and square-root on Intel
https://bugs.webkit.org/show_bug.cgi?id=249275
rdar://103326365

Reviewed by Yusuke Suzuki.

Adds support for missing floating-point opcodes to our WebAssembly SIMD support on Intel, with some
fixes to bugs that came up running the spec tests.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::compareFloatingPointVectorUnordered):
(JSC::MacroAssemblerX86_64::vectorMax):
(JSC::MacroAssemblerX86_64::vectorMin):
(JSC::MacroAssemblerX86_64::vectorPmin):
(JSC::MacroAssemblerX86_64::vectorPmax):
(JSC::MacroAssemblerX86_64::vectorTruncSatSignedFloat64):
(JSC::MacroAssemblerX86_64::vectorSqrt):
(JSC::MacroAssemblerX86_64::vectorSshr8):
(JSC::MacroAssemblerX86_64::vectorUshr8):
(JSC::MacroAssemblerX86_64::vectorSshr):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vmaxps_rrr):
(JSC::X86Assembler::vmaxpd_rrr):
(JSC::X86Assembler::vminps_rrr):
(JSC::X86Assembler::vminpd_rrr):
(JSC::X86Assembler::vsqrtps_rr):
(JSC::X86Assembler::vsqrtpd_rr):
(JSC::X86Assembler::vminpd_mrr):
(JSC::X86Assembler::vpsrlw_i8rr):
(JSC::X86Assembler::vpsrld_i8rr):
(JSC::X86Assembler::vpsrlq_i8rr):
(JSC::X86Assembler::vpsraw_i8rr):
(JSC::X86Assembler::vpsrad_i8rr):
(JSC::X86Assembler::vcmpeqpd_rrr): Deleted.
(JSC::X86Assembler::vpsraq_rrr): Deleted.
* Source/JavaScriptCore/b3/air/AirLowerMacros.cpp:
(JSC::B3::Air::lowerMacros):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDV_V):
(JSC::Wasm::AirIRGenerator64::addSIMDShift):

Canonical link: https://commits.webkit.org/257880@main
